### PR TITLE
Actions refactor and BitSong Fantoken creation

### PR DESCRIPTION
**Pull request by noahsaso**
HUGE

### CODE REVIEW

**Comment by reecepbcups** [comment]
😩😩😩 big

**Comment by JakeHartnell** [comment]
Hot. ❤️

File: packages/stateful/actions/providers/base.tsx [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+    const coreActions = getCoreActions()
+    const coreActionCategoryMakers = getCoreActionCategoryMakers()
+
+    const actions = coreActions.flatMap((Action) => {
+      // Action constructor throws error for invalid contexts.
+      try {
+        return new Action(options)
+      } catch {
+        return []
```
Comment (nikitatrifan) [inline packages/stateful/actions/providers/base.tsx:71]: Do we ever get errors here? Could capture them with Sentry if it's of any meaning ofc

File: packages/stateful/actions/core/actions/AcceptSubDao/index.tsx [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+
+    if (
+      !isValidBech32Address(address, getChainForChainId(chainId).bech32_prefix)
+    ) {
+      throw new Error('Invalid SubDAO address')
+    }
+
+    // Get the current admin of the subDAO and its accounts.
+    const [subDaoAdmin, subDaoAccounts] = await Promise.all([
+      this.options.queryClient.fetchQuery(
```
Comment (nikitatrifan) [inline packages/stateful/actions/core/actions/AcceptSubDao/index.tsx:100]: 👏🏻

File: packages/stateful/actions/core/actions/AcceptSubDao/index.tsx [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+            }),
+          ]
+        : []),
+    ]
+  }
+
+  async match(messages: ProcessedMessage[]): Promise<ActionMatch> {
+    // First (potentially wrapped) message must be accept admin nomination.
+    if (
+      !objectMatchesStructure(messages[0].decodedMessage, {
```
Comment (nikitatrifan) [inline packages/stateful/actions/core/actions/AcceptSubDao/index.tsx:191]: Nice I also like `matchesMessageDefinition`

**Review by nikitatrifan — COMMENTED** [review]
Badass, this is really sweet! I wish typescript was enough on its own so classes weren’t needed, but it’s still pretty sweet like this. If this is to scale, the structure feels way more predictable from a dev’s POV

**Pull request merged**